File: apps/api/modules/Shared/src/Domain/Connectivity/Observation.php
<?php

declare(strict_types=1);

namespace EruoFood\Shared\Domain\Connectivity;

use DateTimeImmutable;

/**
 * A value, when it was observed, and how much that lets you trust it.
 *
 * ## Why the envelope rather than an extra field
 *
 * A response that returns `{"lat": ..., "lng": ...}` and separately
 * `{"location_updated_at": ...}` invites a client to read the first and ignore
 * the second. Binding them into one object means a consumer cannot get the
 * value without having been handed its age — the API shape does the reminding.
 *
 * ## Absence is a first-class answer
 *
 * `Observation::unavailable()` carries no value and says why. A rider whose
 * phone has been off for an hour has no position, and that is *information*:
 * it is different from a position we have not fetched, and different again from
 * a position at the restaurant. Returning null for all three loses the
 * distinction that operations needs.
 *
 * @template T
 */
final readonly class Observation
{
    /**
     * @param T|null $value
     */
    private function __construct(
        public mixed $value,
        public ?DateTimeImmutable $observedAt,
        public FreshnessState $freshness,
        public ?int $ageSeconds,
        public ?string $note,
    ) {
    }

    /**
     * Something we observed, dated, with its freshness already judged.
     *
     * @template V
     *
     * @param V $value
     * @return self<V>
     */
    public static function of(
        mixed $value,
        DateTimeImmutable $observedAt,
        DateTimeImmutable $now,
        FreshnessPolicy $policy,
    ): self {
        $age = max(0, $now->getTimestamp() - $observedAt->getTimestamp());

        return new self($value, $observedAt, $policy->judge($age), $age, null);
    }

    /**
     * Nothing to report, and the reason.
     *
     * @return self<null>
     */
    public static function unavailable(FreshnessState $freshness, ?string $note = null): self
    {
        return new self(null, null, $freshness, null, $note);
    }

    /**
     * A value we hold but cannot date.
     *
     * Deliberately `StaleUnknown` rather than `Online`. Data with no timestamp
     * is data whose age nobody has established, and defaulting that to "current"
     * is how stale information gets presented as live.
     *
     * @template V
     *
     * @param V $value
     * @return self<V>
     */
    public static function undated(mixed $value, ?string $note = null): self
    {
        return new self($value, null, FreshnessState::StaleUnknown, null, $note);
    }

    public function hasValue(): bool
    {
        return $this->value !== null;
    }

    /** Whether a dispatch or pricing decision may rest on this. */
    public function isSafeToActOn(): bool
    {
        return $this->hasValue() && $this->freshness->isSafeToActOn();
    }

    /**
     * The wire shape.
     *
     * `observed_at` and `freshness` are always present, including when there is
     * no value — a client must be able to tell "we have nothing" from "we did
     * not look".
     *
     * @return array{value: mixed, observed_at: string|null, freshness: string, age_seconds: int|null, note: string|null}
     */
    public function toArray(): array
    {
        return [
            'value' => $this->value,
            'observed_at' => $this->observedAt?->format(DATE_ATOM),
            'freshness' => $this->freshness->value,
            'age_seconds' => $this->ageSeconds,
            'note' => $this->note,
        ];
    }
}

File: apps/mobile/lib/core/resilience/observation.dart
import 'freshness.dart';

/// A value, when the server observed it, and how much that lets you trust it.
///
/// The client mirror of the server's `Observation` envelope. Binding the value
/// to its age means a widget cannot render the value without having been handed
/// the caveat — the type does the reminding, rather than a convention nobody
/// remembers at 5pm.
class Observation<T> {
  const Observation({
    required this.value,
    required this.observedAt,
    required this.freshness,
    this.ageSeconds,
    this.note,
  });

  /// Parse the server envelope.
  ///
  /// Anything missing or unrecognised degrades to [FreshnessState.staleUnknown]
  /// rather than defaulting to fresh. A client talking to an older server, or a
  /// response that lost a field, must not start claiming data is live.
  factory Observation.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic) parseValue,
  ) {
    final raw = json['value'];
    final observedAtRaw = json['observed_at'];

    return Observation<T>(
      value: raw == null ? null : parseValue(raw),
      observedAt: observedAtRaw is String
          ? DateTime.tryParse(observedAtRaw)?.toUtc()
          : null,
      freshness: FreshnessState.fromWire(json['freshness'] as String?),
      ageSeconds: json['age_seconds'] as int?,
      note: json['note'] as String?,
    );
  }

  /// Something held locally whose age nobody has established.
  ///
  /// Cached data restored after a restart lands here. It is genuinely useful —
  /// a customer seeing their last known order beats an empty screen — but it is
  /// not current and must never be drawn as though it were.
  factory Observation.cached(T value, {String? note}) => Observation<T>(
        value: value,
        observedAt: null,
        freshness: FreshnessState.staleUnknown,
        note: note ?? 'Restored from local cache; age unknown.',
      );

  /// Nothing to show, and why.
  factory Observation.unavailable({
    FreshnessState freshness = FreshnessState.offline,
    String? note,
  }) =>
      Observation<T>(
          value: null, observedAt: null, freshness: freshness, note: note);

  final T? value;

  /// Always UTC. The device clock is never authoritative — see
  /// `server_time` on the reconciliation response.
  final DateTime? observedAt;

  final FreshnessState freshness;
  final int? ageSeconds;
  final String? note;

  bool get hasValue => value != null;

  /// Whether the UI may present this without an age or a caveat.
  bool get mayPresentAsLive => hasValue && freshness.mayPresentAsLive;

  /// Whether it is worth putting on screen at all, labelled.
  bool get isWorthShowing => hasValue && freshness.isUsableWithCaveat;
}
